### ISSUE TITLE
Support expandable fields

### ISSUE THREAD

**Issue by application-developer-DA** [wyyerd/stripe-rs#56]
I find the idea of introducing `Expand` type on [expandable](https://stripe.com/docs/api/expanding_objects) fields nice, currently a very [similar crate](https://github.com/bundleofbytes/libstripe) provides this feature. `libstripe` seem to have a good way of handling such abstractions. Basically `Expand` is an `enum` which either contains an id or the expanded type `T`. I think it may simplify the usage of the API and speed up the things. Also I've noticed that `*Params` structures miss the `expand` field.

Are there any plans for supporting expanding objects?

**Comment by kestred** [comment]
This seems like a good way to implement expanding objects.   It can definitely be considered planned for the crate.

I'd probably implement it with a few changes for the expandable type.

We would want an extra `Id` parameter to support strongly-typed ids;
additionally I'd prefer the `Box` be built in to the enum rather than part
of each field declaration:

```rust
enum Expandable<Id, Obj> {
    Id(Id),
    Object(Box<Obj>),
}

struct Invoice {
   customer: Expandable<String, Customer>
   // ... other fields
}
```

Generally, we've been attempting to support well-typed ids, so

**Comment by application-developer-DA** [comment]
Sounds good. Is `Box` required though? `libstripe` does not have any boxes at this place and it works fine. I may assume that you expect `Box` to be useful here to avoid possible recursive structure definitions (which would result in the compile time error).

I think it would also be nice to have some helper functions on `Expandable` like the ones I added to `libstripe` recently (https://github.com/bundleofbytes/libstripe/pull/4).

NB: Actually, I think that it would probably make sense to combine efforts and build one good crate for accessing stripe API, I mentioned it yesterday https://github.com/bundleofbytes/libstripe/issues/3 . Both crates seem to have very similar abstractions and architecture. I actually thought to use `stripe-rs` for the production project I'm working on, since `stripe-rs` has more contributors, well documented and slightly more people use it. But I had to switch to `libstripe` as it simply supports more required methods for my particular use case (the bindings seem to be more "up-to-date" for subscription-related functionality) and also due to the good support of the expanding objects. The only problem I see is that `libstripe` requires some refactoring at some places.

To summarize, I think `libstripe` has the following advantages over `stripe-rs`:

* Expanding objects
* More up-to-date state for some of the APIs
* Strongly-typed enumerations instead of `String` types for fields that only take a limited set of values

However `stripe-rs` has slightly bigger team of contributors, better documentation and tests and also what I like is that parameters to the requests are strongly typed (in `libstripe` you can pass any `Deserializable` object into parameter to `XXX_create(client, params)` and for some requests I have to send a dummy `()` just to make it compilable).

If someone could "merge" both crates into one, I think we would have the one good crate for Stripe API :)

**Comment by kestred** [comment]
> Sounds good. Is `Box` required though? `libstripe` does not have any boxes at this place and it works fine. I may assume that you expect `Box` to be useful here to avoid possible recursive structure definitions (which would result in the compile time error).

That crate does use `Box<T>` (see https://github.com/bundleofbytes/libstripe/blob/master/src/resources/orders/order.rs#L21 for example).

My point is actually exactly that we'd prefer to _avoid_ using a box if we aren't using `expand`, which is why I'd rather it be built in to the enum, so that it is only necessary for expanded values rather than on every field that can be expanded.

> If someone could "merge" both crates into one, I think we would have the one good crate for Stripe API :)

Of course for my benefit, I would've preferred that @dariusc93 open a pull request so we can all work together to get to an idiomatic-feeling rust stripe api, but of course there is _a lot_ of noise when you search for a library on crates.io: https://crates.io/search?q=stripe.

**Comment by application-developer-DA** [comment]
Ah, got it, the idea with `Box` makes sense then (I did not know that `libstripe` uses `Box` at _some_ places).

> Of course, I prefer that we all work together to get to an idiomatic-feeling rust stripe api, but of course there is a lot of noise when you search for a library on crates.io: https://crates.io/search?q=stripe.

You're right. I analyzed almost every single crate when I was trying to decide which one to use, but it seems that there are only 2 crates which are still "actively developed": `stripe-rs` (a.k.a. `stripe-rust` on crates.io) and `libstripe`, the rest seem to be obsolete and dead.

**Comment by dariusc93** [comment]
I do agree that `Box` shouldnt be used everywhere especially if someone isnt using `expand`. I do have a local branch with `Box` being used in just the enum and not everywhere, but kept it as is for the time being. Mainly for some internal testing and debugging I was doing. I will probably rebase and push it out later. 

I also do agree that we all should work together and have one crate than to have multiple crates with people not sure what to use or choose. As I mentioned in the issue @application-developer-DA created, while I dont mind merging (or even doing a PR), some of the changes I do may not be suited for merging just yet. I finish on cleaning up the crate, I do want to implement something to make it easier for maintaining the crate when stripe silently updates their documentation (which they have been doing for awhile now).

**Comment by application-developer-DA** [comment]
@kestred BTW, wouldn't it be simpler to replace the source code parts from `stripe-rs` by `libstripe` (doing refactoring in parallel) and then fill the gaps (missing docs etc)? I've tried to do changes in libraries for the product I'm working on, but it turned out that making changes in `libstripe` is much simpler as the APIs are more up-to-date, whereas in `stripe-rs` some APIs are either not complete or implement the deprecated API. I think copy-pasting / replacing the parts of `stripe-rs` by `libstripe` will be simpler than the other way around. 

Also `stripe-rs` is more popular and better documented, so I think that the people who don't investigate differences (like I did) will "by default" choose `stripe-rs`.

**Comment by kestred** [comment]
@application-developer-DA - I'm just planning to autogenerate the API https://github.com/wyyerd/stripe-rs/pull/60

**Comment by kestred** [comment]
Implemented as part of #60.
